src/proofs/ProofSteps.h

```diff
@@ -0,0 +1,147 @@
+
+    std::vector<int> stepsToReuse = {-1, -1, -1};
+
+    std::shared_ptr<Term> implicationRHS;
+    std::shared_ptr<Term> implicationLHS;
+    std::shared_ptr<Term> currTerm;
+    std::vector<int> modusPonensSteps;    // Modus Ponens Steps to derive the next node
+    std::vector<int> simplificationSteps; // Simplification steps for final resolution of LHS
+
+    std::shared_ptr<Term> originalLhsPrimaryBranch;
+
+    // Visitors
+    SimplifyLocatorVisitor simplifyLocatorVisitor;
+    OperateVisitor operateVisitor;
+    SimplifyHelperVisitor helperVisitor;
+    NonLinearVisitor nonLinearVisitor;
+    NegatedAndVisitor negatedAndVisitor;
+    OperateLetTermVisitor operateLetTermVisitor;
+    LetLocatorVisitor letLocatorVisitor;
+    RemoveUnusedVisitor removeUnusedVisitor;
+
+public:
+    StepHandler(InvalidityWitness::Derivation derivation, std::vector<std::shared_ptr<Term>> originalAssertions,
+                Normalizer::Equalities const & normalizingEqualities, std::ostream & out, Logic & logic,
+                ChcDirectedHyperGraph originalGraph)
+        : derivation(std::move(derivation)), originalAssertions(std::move(originalAssertions)),
+          normalizingEqualities(std::move(normalizingEqualities)), originalGraph(std::move(originalGraph)), out(out),
+          logic(logic) {}
+
+    std::vector<std::pair<std::string, std::string>> getInstPairs(int it, vec<Normalizer::Equality> const & stepNormEq);
+    static std::vector<std::shared_ptr<Term>> packClause(const std::shared_ptr<Term> & term);
+    static std::vector<std::shared_ptr<Term>> packClause(const std::shared_ptr<Term> & term1,
+                                                         const std::shared_ptr<Term> & term2);
+    void buildAletheProof();
+    void buildIntermediateProof();
+
+    bool requiresCong();
+
+    void instantiationSteps(int i);
+    void assumptionSteps();
+    void noCongRequiredSteps(std::vector<int> requiredMP, int implicationStep,
+                             const std::shared_ptr<Term> & renamedImpLHS);
+    void notLhsPrimaryBranchSteps(const std::shared_ptr<Term> & simplification);
+    void conjunctionSimplification(std::vector<int> requiredMP, const std::shared_ptr<Term> & simplification,
+                                   const std::shared_ptr<Term> & termToSimplify, std::string simplificationRule,
+                                   int implicationStep, std::shared_ptr<Term> renamedImpLHS, bool cong = true);
+
+    int stepReusage(std::shared_ptr<Term> term);
+
+    void registerObserver(Observer * observer) { observers.push_back(observer); }
+
+    void deRegisterObserver(Observer * observer) {
+        for (int i = 0; i < observers.size(); i++) {
+            if (observer == observers[i]) { observers.erase(observers.begin() + i); }
+        }
+    }
+
+    void notifyObservers(const Step & step) {
+        for (Observer * observer : observers) { // notify all observers
+            observer->update(const_cast<Step &>(step));
```

Comment:
Do not use `const_cast` here.
`update` method  of `Observer` should take `Step const &` as parameter.